openedx/core/djangoapps/agreements/api.py

```diff
@@ -68,3 +73,180 @@ def get_integrity_signatures_for_course(course_id):
     """
     course_key = CourseKey.from_string(course_id)
     return IntegritySignature.objects.filter(course_key=course_key)
+
+
+def create_lti_pii_signature(username, course_id, lti_tools):
+    """
+    Creates an lti pii tool signature. If the signature already exist, do not create a new one.
+
+    Arguments:
+        * course_key (str)
+        * lti_tools (dict)
+        * lti_tools_hash (int)
+    Returns:
+        * An LTIPIISignature, or None if a signature already exists.
+    """
+    user = User.objects.get(username=username)
+    course_key = CourseKey.from_string(course_id)
+    lti_tools_hash = hash(str(lti_tools))
+
+    # if user and course exists, update, otherwise create a new signature
+    try:
+        LTIPIISignature.objects.get(user=user, course_key=course_key)
```

Comment:
Using the `ObjectDoesNotExist` exception in a `try...catch` block will "[will catch DoesNotExist exceptions for all models](https://docs.djangoproject.com/en/4.2/ref/exceptions/#objectdoesnotexist)". This could be a problem because we might catch exceptions for other models that we don't expect, and then the exception is buried and may not be handled appropriately. I think it would be better to catch the exception that you anticipate and want to handle, which is `LTIPIISignature.DoesNotExist`. You can see the documentation for this exception [here](https://docs.djangoproject.com/en/4.1/ref/models/class/#django.db.models.Model.DoesNotExist).

I see this `ObjectDoesNotExist` class in a few places in this file.

Reply:
Ohh, I was wondering how to throw a `DoesNotExist` exception for specific models. I just resorted to `ObjectDoesNotExist`. I will correct the rest of them as well.